File: 36. Valid Sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        '''
        this is a one pass check, key point is build a hashmap to store values that checked before
        the trick of the blocks is mark blocks as 0-8 and access them by (i//3)*3+j//3
        '''
        block = [{} for i in range(9)]
        row = [{} for i in range(9)]
        column = [{} for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j].isdigit():
                    if board[i][j] in row[i] or board[i][j] in column[j] or board[i][j] in block[(i//3)*3+j//3]:
                        return False
                    else:
                        row[i][board[i][j]] = 1
                        column[j][board[i][j]] = 1
                        block[(i//3)*3+j//3][board[i][j]] = 1
        return True

File: 36. Valid Sudoku.java
class Solution {
    private char[][] board;
    public boolean isValidSudoku(char[][] board) {
        this.board = board;
        boolean isValid = true;
        for (int i=0; i<9; i++) {
            isValid = isValid && checkRow(i) && checkCol(i) && checkBlock(i);
        }
        return isValid;
    }
    private boolean checkRow(int row) {
        boolean[] check = new boolean[9];
        for (int i=0; i<9; i++) {
            if (Character.isDigit(board[row][i])) {
                int t = board[row][i]-'1';
                if (check[t]) {
                    return false;
                } else {
                    check[t] = true;
                }
            }
        }
        return true;
    }
    private boolean checkCol(int col) {
        boolean[] check = new boolean[9];
        for (int i=0; i<9; i++) {
            if (Character.isDigit(board[i][col])) {
                int t = board[i][col]-'1';
                if (check[t]) {
                    return false;
                } else {
                    check[t] = true;
                }
            }
        }
        return true;
    }
    private boolean checkBlock(int pos) {
        boolean[] check = new boolean[9];
        int r = pos/3*3, c = pos%3*3;
        for (int i=0; i<3; i++) {
            for (int j=0; j<3; j++) {
                if (Character.isDigit(board[i+r][j+c])) {
                    int t = board[i+r][j+c]-'1';
                    if (check[t]) {
                        return false;
                    } else {
                        check[t] = true;
                    }
                }
            }
        }
        return true;
    }
}